### ISSUE TITLE
[PSaaS TASK] Test dataset evaluation

### ISSUE THREAD

**Issue by RobBryce** [PSaaS-Developers/Project_issues#77]
REDMine #1175 covered the collection of specific datasets for fires in the 2021 season, to help with V6 to V7 validation.

We are entering the 2022 season using V7, but without any work on this validation.

I ask that we have a discussion to formulate a plan on what to do with this dataset (so it doesn't end up forgotten), and when to do it, so that the project can benefit from this prior work as best possible.  That may mean defining a test methodology now and testing later, or validating before the fire season truly begins.

**Comment by nealmcloughlin** [comment]
I agree Rob. We need to start with defining a test methodology. I'd like to know what you are thinking about with respect to comparing V6 and V7 fire datasets. What are the metrics and criteria for pass/fail? I was thinking we could use the overlay metrics Brett used in his MSc for the comparison where V6 outputs are considered "the truth". However, this would be comparing raster burn grids and not the more detailed fire perimeters. However, I'm honest not worried about minor differences that don't show up in a raster burn grid. Thoughts?

**Comment by RobBryce** [comment]
The test datasets presumably cover a variety of optional inputs.  This is important for breadth of the test using 2021 data.  This ranges from different time zones, to different LUT's, to different grid layers and possibly vector data.

The test datasets presumably also have varying configurations for PSaaS/Prometheus options, which is also important for breadth of the test.  If the datasets don't include FGMJ's, then we should define which options we should test, that we expect to use regularly (and maybe a few that are irregular).

Although I think it's expected to assume that V6 is the "truth", it may not be.  As Neal and Franco have mentioned, all models are wrong.  And if we see something anomalous, then we have to find out why and V7 validation has already fixed the odd bug in V6.

We have also compared V6 and V7 in other cases.  For short duration or small outputs, visually there may be no difference for a while, but very small differences to 4 or 5 decimal places may introduce variations in outputs that appear over time.  For this reason, we cannot simply compare the last perimeters or fire extents, we need to compare them on a variety of time steps.  This will let us observe if there are significant differences in (e.g.) fire ROSv, rather than incredibly small differences in ROSv that finally let one vertex hit a different cell.  As soon as a single vertex lands on a different cell, changes will propagate quickly, since that one vertex will change direction and magnitude of growth, and thus impact neighbouring PDE calculations and where new vertices are introduced.

I think we need to generate complex FGMJ's that contain a variety of scenarios that will run different configurations that are of interest to us.  And each scenario will have a variety of outputs that can include gridded statistics and per-vector statistics (where possible).  We have a tool that lets us automate execution of a variety of scenario configurations, so we should use it (but we may need a tool to automate the building of the scenario configurations, that's something different).  That will let us visually observe outputs, but also let us examine differences that appear to be errors vs. what can be propagated from rounding errors and the nature of a spatially explicit model based on sampling heterogeneous landscape (that's enough buzzwords for one sentence!)

I'm less concerned about differences in outputs, as I am why the differences are occurring.  Many can be explained and accepted, but we need to watch for ones that appear erroneous.  But I'm not yet sure how we can automate the comparison of V6 vs V7 outputs.  If files are precisely the same to N decimal places (if exported as ASCII) then we have a good place to start.  There may be GIS tools to let us statistically compare polygon shapes, but that may not uncover what we should be looking for.

We also need to generate a series of output files for gridded statistics.  We introduced this for New Zealand validation and found a fixed a variety of differences.  If we run the same tests on this dataset, then we can be confident that all inputs to the FireEngine DLL match.  If those match (and they should), then we only need to look for what may be rounding errors vs. coding errors in FireEngine, which does the actual propagation.

I know I'm all over the map in this post, but it's a big topic.  In conclusion:
* we need breadth of testing different options, and may need a tool to generate scenario settings
* we have a tool (PSaaS) that can run over (e.g.) 400 scenario settings in FGMJ, and generate all outputs we need
* we can automate comparisons of a large set of gridded exports from a simulation (e.g. HROS, BFI, RAZ, etc. over the entire grid at a given time)
* if those comparisons fail, we need to investigate.  If those comparisons pass, we may have to semi-automate and spend time manually comparing output results from FireEngine in the form of raster outputs and vector outputs

With some forethought, we can also produce other stat's like how V6 compares to V7 in performance when the simulations produce identical results.  And ensure that V7-Windows produces the same outputs as V7-Linux.

I feel we have a good level of confidence in V7, and we do expect differences in outputs (just like we saw differences from 32-bit to 64-bit targets).  But I created this post in case we wanted to do some diligence before the season, not during the season.

**Comment by spydmobile** [comment]
@BadgerOnABike we need to sort this ticket into multiple or decode how we want to move forward.

**Comment by BadgerOnABike** [comment]
We're talking about a few things here and I think there is a bit of conflation of issues going on.

1) we need to test v6 vs v7 for discrepancy
- 100%, none of this data needs to be real, but all settings perms/combs should be tested, this should be 100% automatic and it can just rotate through all the scenarios and do a flat comparison.
2) How are 6 and 7 doing compared to reality and each other.
- This is a fundamentally different 6 to 7 comparison as they aren't actually compared to one another, rather their metrics are compared based on performance in reality.

In my eyes we need to do 1 before we can consider 2, because it doesn't matter how they are doing in reality if they are significantly different from one another (and shouldn't be). I add the shouldn't be part in brackets because, should they be fundamentally different? Maybe there are some significant differences? 

Anyway, we need that swath of unit tests to understand the actual model level differences before we worry about complex real world fires.

These are far more simple, 1 fuel, flat terrain, some set of options and burn (then ratchet through all the options)
Introduce complex fuels and terrain, rinse, repeat.

This will require multiple tickets @spydmobile. Perhaps we build a framework here with some checkboxes and then flesh them out.

*Unit Testing*

- [ ] - Define all options for testing
- [ ] - Define all reasonable perms/combs
- [ ] - Test unreasonable perms/combs to ensure proper failure
- [ ] - Test all with 1 fuel and flat terrain
- [ ] - Test all with 1 fuel and complex terrain
- [ ] - Test all with 1 fuel and complex terrain and wind grids
- [ ] - Test all with complex fuel and complex terrain
- [ ] - Test all with complex fuel and complex terrain and wind grids

Something like this. The actual testing would be 1 ticket with many tests embedded.

*Reworld Assessment*

- [ ] - Use at least 25 fires that have good ground truth chronologies
- [ ] - Establish an FGMJ to set the project up in both v6 and v7 identically
- [ ] - Define metrics to describe accuracy against the real world data
- [ ] - Compare metrics from 6 to 7
